File: swift-app/Utils/Supabase.swift
//
//  Supabase.swift
//  MeetJournal
//
//  Created by Maddisen Mohnsen on 12/25/25.
//

import Foundation
import Supabase
import Clerk

let supabaseURL = Bundle.main.object(forInfoDictionaryKey: "SUPABASE_URL") as! String
let supabaseKey = Bundle.main.object(forInfoDictionaryKey: "SUPABASE_KEY") as! String

let supabase: SupabaseClient = {
    let options = SupabaseClientOptions(
        auth: .init(
            accessToken: {
                guard let session = Clerk.shared.session else {
                    return ""
                }
                
                let semaphore = DispatchSemaphore(value: 0)
                var jwtToken = ""
                
                Task {
                    do {
                        if let token = try await session.getToken() {
                            jwtToken = token.jwt
                        }
                    } catch {
                        print("Error getting Clerk token: \(error)")
                    }
                    semaphore.signal()
                }
                
                semaphore.wait()
                return jwtToken
            }
        )
    )
    
    return SupabaseClient(
        supabaseURL: URL(string: supabaseURL)!,
        supabaseKey: supabaseKey,
        options: options
    )
}()

func getSupabaseClient() -> SupabaseClient {
    return supabase
}

File: services/supabase.ts
import { createClient, type SupabaseClient } from '@supabase/supabase-js';

const supabaseUrl = process.env.EXPO_PUBLIC_SUPABASE_URL ?? '';
const supabaseKey = process.env.EXPO_PUBLIC_SUPABASE_KEY ?? '';

export const createSupabaseClient = (getToken?: () => Promise<string | null>): SupabaseClient => {
  if (!supabaseUrl || !supabaseKey) {
    throw new Error('Supabase URL/Key missing. Set EXPO_PUBLIC_SUPABASE_URL and EXPO_PUBLIC_SUPABASE_KEY.');
  }
  return createClient(supabaseUrl, supabaseKey, {
    global: {
      fetch: async (url, init = {}) => {
        const headers = new Headers(init.headers ?? {});
        if (getToken) {
          const token = await getToken();
          if (token) {
            headers.set('Authorization', `Bearer ${token}`);
          }
        }
        return fetch(url, { ...init, headers });
      },
    },
  });
};

export const supabase = supabaseUrl && supabaseKey ? createSupabaseClient() : null;
